### PR TITLE
debug build support for ISTGT

### PR DESCRIPTION
**Pull request by mynktl**
Changes : 

- Added debug build support
    - To enable debug build, configure with `--enable-debug` argument.
    - To enable replication support, configure with `--enable-replication` argument
    - To enable debug logging(tracedump) for ISTGT, configure with `--enable-tracedump`. (before this, it was `--enable-debug`. now changed to `--enable-tracedump`)

- Changes in travis file to use debug build in travis-build

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

### CODE REVIEW

File: src/data_conn.c [hunk 1]
```diff
@@ -313,6 +319,7 @@ read_cmd(replica_t *r)
 
 	switch(state) {
 		case READ_IO_RESP_HDR:
+			VERIFY(r->io_read < sizeof(zvol_io_hdr_t));
```
Comment (satbirchhikara) [inline src/data_conn.c]: Please add default in switch and add apppropriate verify string.

File: src/replication.c [hunk 2]
```diff
@@ -1033,9 +1045,12 @@ write_io_data(replica_t *replica, io_event_t *wevent)
 	int donecount = 0;
 	(void)replica;
 
+	VERIFY((*state == WRITE_IO_SEND_HDR) || (*state == WRITE_IO_SEND_DATA));
```
Comment (satbirchhikara) [inline src/replication.c]: Add default in switch and add verify under default with a proper string.

File: src/replication.c [hunk 3]
```diff
@@ -1080,10 +1095,13 @@ read_io_resp(spec_t *spec, replica_t *replica, io_event_t *revent, mgmt_cmd_t *m
 	int rc = 0;
 	int donecount = 0;
 
+	VERIFY((*state == READ_IO_RESP_HDR) || (*state == READ_IO_RESP_DATA));
```
Comment (satbirchhikara) [inline src/replication.c]: Same goes here too.

File: src/replication.c [hunk 4]
```diff
@@ -1234,10 +1251,7 @@ free_replica(replica_t *r)
 void
 close_fd(int epollfd, int fd)
 {
-	if (epoll_ctl(epollfd, EPOLL_CTL_DEL, fd, NULL) == -1) {
-		REPLICA_ERRLOG("epoll error for fd(%d) err(%d)\n", fd, errno);
-		return;
-	}
+	ASSERT0(epoll_ctl(epollfd, EPOLL_CTL_DEL, fd, NULL));
```
Comment (satbirchhikara) [inline src/replication.c]: Take care of it, macro will replace this code with _DO_NOTHING_ in production build.

**Review by satbirchhikara — COMMENTED** [review]
Hi Mayank,

Give some minor comments. Please have a look and let me know if you need discussion. General comment is that 
add default case in all switch statements if possible for you and add verify. 

Thanks & Regards,
Satbir Singh

File: src/data_conn.c (hunk 1 above)
Comment (mynktl) [reply src/data_conn.c]: ok

File: src/replication.c (hunk 2 above)
Comment (mynktl) [reply src/replication.c]: ok

File: src/replication.c (hunk 3 above)
Comment (mynktl) [reply src/replication.c]: ok

File: src/replication.c (hunk 4 above)
Comment (mynktl) [reply src/replication.c]: ok. Thanks.

**Comment by mynktl** [comment]
Hi @satbirchhikara 
I have updated PR with review comments and also with a fix for a segfault in ISTGT seen in recent PR.
Please have a look and let me know if any changes required.

**Review by satbirchhikara — APPROVED** [review]
LGTM

File: src/replication.c [hunk 5]
```diff
@@ -1925,23 +1948,24 @@ replicate(ISTGT_LU_DISK *spec, ISTGT_LU_CMD_Ptr cmd, uint64_t offset, uint64_t n
...
-				    errno, rcomm_cmd->opcode);
+				fprintf(stderr, "last errno(%d) "
+				    "opcode(%d)\n", errno, rcomm_cmd->opcode);
 
 			MTX_LOCK(&spec->rq_mtx);
 			TAILQ_REMOVE(&spec->rcommon_waitq, rcomm_cmd, wait_cmd_next);
 			MTX_UNLOCK(&spec->rq_mtx);
 
+			put_to_mempool(&spec->rcommon_deadlist, rcomm_cmd);
+			MTX_UNLOCK(rcomm_cmd->mutex);
```
Comment (vishnuitta) [inline src/replication.c:1968]: can this lock sequence lead to deadlock?
And, why do we need lock to add to mempool?

**Review by vishnuitta — APPROVED** [review]
lock time in replicate can be reduced further. Please look into it in next PR if it can be reduced.

**Pull request merged**